File: protocol/runtime/typescript/ingress.ts
export type IngressErrorCode =
  | "PAYLOAD_TOO_LARGE"
  | "INVALID_UTF8"
  | "MALFORMED_JSON"
  | "DUPLICATE_JSON_MEMBER"
  | "JSON_LIMIT_EXCEEDED"
  | "UNSAFE_JSON_NUMBER";

export class IngressError extends Error {
  readonly code: IngressErrorCode;

  constructor(code: IngressErrorCode, message: string) {
    super(`${code}: ${message}`);
    this.code = code;
    Object.setPrototypeOf(this, IngressError.prototype);
  }
}

const MAX_BODY_BYTES = 1048576;
const MAX_DEPTH = 64;
const MAX_TOKENS = 100000;
const MAX_MEMBERS = 4096;
const MAX_ITEMS = 10000;
const MIN_SAFE_INT = -9007199254740991;
const MAX_SAFE_INT = 9007199254740991;

export function parseRawIngress(rawBytes: Uint8Array): unknown {
  if (rawBytes.length > MAX_BODY_BYTES) {
    throw new IngressError("PAYLOAD_TOO_LARGE", `body length ${rawBytes.length} exceeds max ${MAX_BODY_BYTES}`);
  }

  let text: string;
  try {
    const decoder = new TextDecoder("utf-8", { fatal: true });
    text = decoder.decode(rawBytes);
  } catch {
    throw new IngressError("INVALID_UTF8", "raw request body is not valid UTF-8");
  }

  let pos = 0;
  let tokenCount = 0;

  function countToken(): void {
    tokenCount++;
    if (tokenCount > MAX_TOKENS) {
      throw new IngressError("JSON_LIMIT_EXCEEDED", `token count exceeds maximum ${MAX_TOKENS}`);
    }
  }

  function skipWhitespace(): void {
    while (pos < text.length) {
      const ch = text.charCodeAt(pos);
      if (ch === 0x20 || ch === 0x09 || ch === 0x0a || ch === 0x0d) {
        pos++;
      } else {
        break;
      }
    }
  }

  function parseValue(depth: number): unknown {
    if (depth > MAX_DEPTH) {
      throw new IngressError("JSON_LIMIT_EXCEEDED", `nesting depth exceeds maximum ${MAX_DEPTH}`);
    }
    skipWhitespace();
    if (pos >= text.length) {
      throw new IngressError("MALFORMED_JSON", "unexpected end of JSON input");
    }

    const ch = text.charCodeAt(pos);
    if (ch === 123 /* { */) {
      return parseObject(depth);
    } else if (ch === 91 /* [ */) {
      return parseArray(depth);
    } else if (ch === 34 /* " */) {
      return parseString();
    } else if (ch === 116 /* t */) {
      return parseLiteral("true", true);
    } else if (ch === 102 /* f */) {
      return parseLiteral("false", false);
    } else if (ch === 110 /* n */) {
      return parseLiteral("null", null);
    } else if (ch === 45 /* - */ || (ch >= 48 && ch <= 57)) {
      return parseNumber();
    } else {
      throw new IngressError("MALFORMED_JSON", "unexpected character");
    }
  }

  function parseLiteral(expected: string, value: unknown): unknown {
    countToken();
    if (text.startsWith(expected, pos)) {
      pos += expected.length;
      return value;
    }
    throw new IngressError("MALFORMED_JSON", "invalid literal");
  }

  function parseString(): string {
    countToken();
    if (text.charCodeAt(pos) !== 34) {
      throw new IngressError("MALFORMED_JSON", "expected string quote");
    }
    pos++;
    let result = "";
    while (pos < text.length) {
      const ch = text.charCodeAt(pos);
      if (ch === 34 /* " */) {
        pos++;
        return result;
      }
      if (ch === 92 /* \ */) {
        pos++;
        if (pos >= text.length) {
          throw new IngressError("MALFORMED_JSON", "unterminated escape sequence");
        }
        const esc = text.charCodeAt(pos);
        pos++;
        switch (esc) {
          case 34: result += '"'; break;
          case 92: result += '\\'; break;
          case 47: result += '/'; break;
          case 98: result += '\b'; break;
          case 102: result += '\f'; break;
          case 110: result += '\n'; break;
          case 114: result += '\r'; break;
          case 116: result += '\t'; break;
          case 117: { // \uXXXX
            if (pos + 4 > text.length) {
              throw new IngressError("MALFORMED_JSON", "invalid hex escape sequence");
            }
            const hex = text.slice(pos, pos + 4);
            if (!/^[0-9a-fA-F]{4}$/.test(hex)) {
              throw new IngressError("MALFORMED_JSON", "invalid hex code");
            }
            pos += 4;
            const codeUnit = parseInt(hex, 16);
            if (codeUnit >= 0xd800 && codeUnit <= 0xdbff) {
              if (pos + 6 <= text.length && text.slice(pos, pos + 2) === "\\u") {
                const lowHex = text.slice(pos + 2, pos + 6);
                if (/^[0-9a-fA-F]{4}$/.test(lowHex)) {
                  const lowUnit = parseInt(lowHex, 16);
                  if (lowUnit >= 0xdc00 && lowUnit <= 0xdfff) {
                    pos += 6;
                    const cp = 0x10000 + ((codeUnit - 0xd800) << 10) + (lowUnit - 0xdc00);
                    result += String.fromCodePoint(cp);
                    break;
                  }
                }
              }
              throw new IngressError("MALFORMED_JSON", "escaped lone high surrogate");
            }
            if (codeUnit >= 0xdc00 && codeUnit <= 0xdfff) {
              throw new IngressError("MALFORMED_JSON", "escaped lone low surrogate");
            }
            result += String.fromCharCode(codeUnit);
            break;
          }
          default:
            throw new IngressError("MALFORMED_JSON", "invalid escape character");
        }
      } else if (ch < 0x20) {
        throw new IngressError("MALFORMED_JSON", "unescaped control character in string");
      } else {
        result += text[pos];
        pos++;
      }
    }
    throw new IngressError("MALFORMED_JSON", "unterminated string literal");
  }

  function parseNumber(): number {
    countToken();
    const start = pos;
    let isNeg = false;
    if (text.charCodeAt(pos) === 45 /* - */) {
      isNeg = true;
      pos++;
    }
    if (pos >= text.length) throw new IngressError("MALFORMED_JSON", "invalid number literal");

    const intStart = pos;
    if (text.charCodeAt(pos) === 48 /* 0 */) {
      pos++;
    } else if (text.charCodeAt(pos) >= 49 && text.charCodeAt(pos) <= 57) {
      while (pos < text.length && text.charCodeAt(pos) >= 48 && text.charCodeAt(pos) <= 57) {
        pos++;
      }
    } else {
      throw new IngressError("MALFORMED_JSON", "invalid number format");
    }
    const intDigits = text.slice(intStart, pos);

    let fracDigits = "";
    if (pos < text.length && text.charCodeAt(pos) === 46 /* . */) {
      pos++;
      const fracStart = pos;
      while (pos < text.length && text.charCodeAt(pos) >= 48 && text.charCodeAt(pos) <= 57) {
        pos++;
      }
      if (pos === fracStart) throw new IngressError("MALFORMED_JSON", "invalid number fraction");
      fracDigits = text.slice(fracStart, pos);
    }

    let expSign = 1;
    let expDigits = "";
    if (pos < text.length && (text.charCodeAt(pos) === 101 || text.charCodeAt(pos) === 69) /* e E */) {
      pos++;
      if (pos < text.length && (text.charCodeAt(pos) === 43 || text.charCodeAt(pos) === 45)) {
        if (text.charCodeAt(pos) === 45) expSign = -1;
        pos++;
      }
      const expStart = pos;
      while (pos < text.length && text.charCodeAt(pos) >= 48 && text.charCodeAt(pos) <= 57) {
        pos++;
      }
      if (pos === expStart) throw new IngressError("MALFORMED_JSON", "invalid number exponent");
      expDigits = text.slice(expStart, pos);
    }

    const expVal = expDigits === "" ? 0n : BigInt(expDigits) * BigInt(expSign);
    const digitsStr = intDigits + fracDigits;
    const effectiveExp = expVal - BigInt(fracDigits.length);

    let exactInt: bigint;
    if (effectiveExp >= 0n) {
      exactInt = BigInt(digitsStr) * (10n ** effectiveExp);
    } else {
      const k = -effectiveExp;
      const divisor = 10n ** k;
      const num = BigInt(digitsStr);
      if (num % divisor !== 0n) {
        throw new IngressError("UNSAFE_JSON_NUMBER", "number is not an integer");
      }
      exactInt = num / divisor;
    }

    if (isNeg) {
      exactInt = -exactInt;
    }

    const minSafe = BigInt(MIN_SAFE_INT);
    const maxSafe = BigInt(MAX_SAFE_INT);
    if (exactInt < minSafe || exactInt > maxSafe) {
      throw new IngressError("UNSAFE_JSON_NUMBER", "number exceeds safe integer bounds");
    }

    return Number(exactInt);
  }

  function parseObject(depth: number): Record<string, unknown> {
    countToken();
    pos++;
    const obj: Record<string, unknown> = {};
    const keys = new Set<string>();
    let memberCount = 0;

    skipWhitespace();
    if (pos < text.length && text.charCodeAt(pos) === 125 /* } */) {
      countToken();
      pos++;
      return obj;
    }

    while (pos < text.length) {
      skipWhitespace();
      if (text.charCodeAt(pos) !== 34 /* " */) {
        throw new IngressError("MALFORMED_JSON", "expected member name string");
      }
      const key = parseString();
      if (keys.has(key)) {
        throw new IngressError("DUPLICATE_JSON_MEMBER", "duplicate object member name");
      }
      keys.add(key);
      memberCount++;
      if (memberCount > MAX_MEMBERS) {
        throw new IngressError("JSON_LIMIT_EXCEEDED", `object member count exceeds maximum ${MAX_MEMBERS}`);
      }

      skipWhitespace();
      if (pos >= text.length || text.charCodeAt(pos) !== 58 /* : */) {
        throw new IngressError("MALFORMED_JSON", "expected ':' after member name");
      }
      countToken();
      pos++;

      const val = parseValue(depth + 1);
      obj[key] = val;

      skipWhitespace();
      if (pos >= text.length) {
        throw new IngressError("MALFORMED_JSON", "unexpected end of object");
      }
      const nextChar = text.charCodeAt(pos);
      if (nextChar === 125 /* } */) {
        countToken();
        pos++;
        return obj;
      }
      if (nextChar === 44 /* , */) {
        countToken();
        pos++;
      } else {
        throw new IngressError("MALFORMED_JSON", "expected ',' or '}'");
      }
    }
    throw new IngressError("MALFORMED_JSON", "unterminated object literal");
  }

  function parseArray(depth: number): unknown[] {
    countToken();
    pos++;
    const arr: unknown[] = [];

    skipWhitespace();
    if (pos < text.length && text.charCodeAt(pos) === 93 /* ] */) {
      countToken();
      pos++;
      return arr;
    }

    while (pos < text.length) {
      if (arr.length >= MAX_ITEMS) {
        throw new IngressError("JSON_LIMIT_EXCEEDED", `array item count exceeds maximum ${MAX_ITEMS}`);
      }
      const val = parseValue(depth + 1);
      arr.push(val);

      skipWhitespace();
      if (pos >= text.length) {
        throw new IngressError("MALFORMED_JSON", "unexpected end of array");
      }
      const nextChar = text.charCodeAt(pos);
      if (nextChar === 93 /* ] */) {
        countToken();
        pos++;
        return arr;
      }
      if (nextChar === 44 /* , */) {
        countToken();
        pos++;
      } else {
        throw new IngressError("MALFORMED_JSON", "expected ',' or ']'");
      }
    }
    throw new IngressError("MALFORMED_JSON", "unterminated array literal");
  }

  const rootValue = parseValue(1);
  skipWhitespace();
  if (pos < text.length) {
    throw new IngressError("MALFORMED_JSON", "unexpected trailing content");
  }
  return rootValue;
}

File: protocol/runtime/go/ingress.go
package protocolruntime

import (
	"bytes"
	"fmt"
	"math/big"
	"strconv"
	"unicode/utf8"
)

const (
	MaxBodyBytes     = 1048576
	MaxJSONDepth     = 64
	MaxJSONTokens    = 100000
	MaxObjectMembers = 4096
	MaxArrayItems    = 10000
	MinSafeInteger   = -9007199254740991
	MaxSafeInteger   = 9007199254740991
)

type IngressError struct {
	Code    string
	Message string
}

func (e *IngressError) Error() string {
	return fmt.Sprintf("%s: %s", e.Code, e.Message)
}

func ParseRawIngress(rawBytes []byte) (any, error) {
	if len(rawBytes) > MaxBodyBytes {
		return nil, &IngressError{
			Code:    "PAYLOAD_TOO_LARGE",
			Message: "payload size exceeds maximum allowed limit",
		}
	}

	if !utf8.Valid(rawBytes) {
		return nil, &IngressError{
			Code:    "INVALID_UTF8",
			Message: "raw request body is not valid UTF-8",
		}
	}

	pos := 0
	tokenCount := 0
	text := string(rawBytes)

	countToken := func() error {
		tokenCount++
		if tokenCount > MaxJSONTokens {
			return &IngressError{
				Code:    "JSON_LIMIT_EXCEEDED",
				Message: fmt.Sprintf("token count exceeds maximum %d", MaxJSONTokens),
			}
		}
		return nil
	}

	skipWhitespace := func() {
		for pos < len(text) {
			ch := text[pos]
			if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
				pos++
			} else {
				break
			}
		}
	}

	var parseValue func(depth int) (any, error)
	var parseObject func(depth int) (map[string]any, error)
	var parseArray func(depth int) ([]any, error)
	var parseString func() (string, error)
	var parseNumber func() (int64, error)
	var parseLiteral func(expected string, val any) (any, error)

	parseLiteral = func(expected string, val any) (any, error) {
		if err := countToken(); err != nil {
			return nil, err
		}
		if stringsHasPrefix(text[pos:], expected) {
			pos += len(expected)
			return val, nil
		}
		return nil, &IngressError{Code: "MALFORMED_JSON", Message: "invalid literal"}
	}

	parseString = func() (string, error) {
		if err := countToken(); err != nil {
			return "", err
		}
		if pos >= len(text) || text[pos] != '"' {
			return "", &IngressError{Code: "MALFORMED_JSON", Message: "expected string quote"}
		}
		pos++
		var buf bytes.Buffer
		for pos < len(text) {
			ch := text[pos]
			if ch == '"' {
				pos++
				return buf.String(), nil
			}
			if ch == '\\' {
				pos++
				if pos >= len(text) {
					return "", &IngressError{Code: "MALFORMED_JSON", Message: "unterminated escape sequence"}
				}
				esc := text[pos]
				pos++
				switch esc {
				case '"':
					buf.WriteByte('"')
				case '\\':
					buf.WriteByte('\\')
				case '/':
					buf.WriteByte('/')
				case 'b':
					buf.WriteByte('\b')
				case 'f':
					buf.WriteByte('\f')
				case 'n':
					buf.WriteByte('\n')
				case 'r':
					buf.WriteByte('\r')
				case 't':
					buf.WriteByte('\t')
				case 'u':
					if pos+4 > len(text) {
						return "", &IngressError{Code: "MALFORMED_JSON", Message: "invalid hex escape sequence"}
					}
					hexStr := text[pos : pos+4]
					codeUnit, err := strconv.ParseUint(hexStr, 16, 16)
					if err != nil {
						return "", &IngressError{Code: "MALFORMED_JSON", Message: "invalid hex code"}
					}
					pos += 4
					if codeUnit >= 0xd800 && codeUnit <= 0xdbff {
						if pos+6 <= len(text) && text[pos:pos+2] == "\\u" {
							lowHex := text[pos+2 : pos+6]
							lowUnit, err := strconv.ParseUint(lowHex, 16, 16)
							if err == nil && lowUnit >= 0xdc00 && lowUnit <= 0xdfff {
								pos += 6
								r := rune(0x10000 + ((codeUnit - 0xd800) << 10) + (lowUnit - 0xdc00))
								buf.WriteRune(r)
								break
							}
						}
						return "", &IngressError{Code: "MALFORMED_JSON", Message: "escaped lone high surrogate"}
					}
					if codeUnit >= 0xdc00 && codeUnit <= 0xdfff {
						return "", &IngressError{Code: "MALFORMED_JSON", Message: "escaped lone low surrogate"}
					}
					buf.WriteRune(rune(codeUnit))
				default:
					return "", &IngressError{Code: "MALFORMED_JSON", Message: "invalid escape character"}
				}
			} else if ch < 0x20 {
				return "", &IngressError{Code: "MALFORMED_JSON", Message: "unescaped control character in string"}
			} else {
				buf.WriteByte(ch)
				pos++
			}
		}
		return "", &IngressError{Code: "MALFORMED_JSON", Message: "unterminated string literal"}
	}

	parseNumber = func() (int64, error) {
		if err := countToken(); err != nil {
			return 0, err
		}
		isNeg := false
		if text[pos] == '-' {
			isNeg = true
			pos++
		}
		if pos >= len(text) {
			return 0, &IngressError{Code: "MALFORMED_JSON", Message: "invalid number literal"}
		}
		intStart := pos
		if text[pos] == '0' {
			pos++
		} else if text[pos] >= '1' && text[pos] <= '9' {
			for pos < len(text) && text[pos] >= '0' && text[pos] <= '9' {
				pos++
			}
		} else {
			return 0, &IngressError{Code: "MALFORMED_JSON", Message: "invalid number format"}
		}
		intDigits := text[intStart:pos]

		fracDigits := ""
		if pos < len(text) && text[pos] == '.' {
			pos++
			fracStart := pos
			for pos < len(text) && text[pos] >= '0' && text[pos] <= '9' {
				pos++
			}
			if pos == fracStart {
				return 0, &IngressError{Code: "MALFORMED_JSON", Message: "invalid number fraction"}
			}
			fracDigits = text[fracStart:pos]
		}

		expSign := int64(1)
		expDigits := ""
		if pos < len(text) && (text[pos] == 'e' || text[pos] == 'E') {
			pos++
			if pos < len(text) && (text[pos] == '+' || text[pos] == '-') {
				if text[pos] == '-' {
					expSign = -1
				}
				pos++
			}
			expStart := pos
			for pos < len(text) && text[pos] >= '0' && text[pos] <= '9' {
				pos++
			}
			if pos == expStart {
				return 0, &IngressError{Code: "MALFORMED_JSON", Message: "invalid number exponent"}
			}
			expDigits = text[expStart:pos]
		}

		var expVal int64
		if expDigits != "" {
			parsedExp, err := strconv.ParseInt(expDigits, 10, 64)
			if err != nil {
				return 0, &IngressError{Code: "UNSAFE_JSON_NUMBER", Message: "invalid number exponent"}
			}
			expVal = parsedExp * expSign
		}

		digitsStr := intDigits + fracDigits
		effectiveExp := expVal - int64(len(fracDigits))

		bigInt := new(big.Int)
		if _, ok := bigInt.SetString(digitsStr, 10); !ok {
			return 0, &IngressError{Code: "UNSAFE_JSON_NUMBER", Message: "invalid number digits"}
		}

		if effectiveExp >= 0 {
			if effectiveExp > 10000 {
				return 0, &IngressError{Code: "UNSAFE_JSON_NUMBER", Message: "number exponent too large"}
			}
			mul := new(big.Int).Exp(big.NewInt(10), big.NewInt(effectiveExp), nil)
			bigInt.Mul(bigInt, mul)
		} else {
			k := -effectiveExp
			if k > 10000 {
				return 0, &IngressError{Code: "UNSAFE_JSON_NUMBER", Message: "number exponent too small"}
			}
			div := new(big.Int).Exp(big.NewInt(10), big.NewInt(k), nil)
			rem := new(big.Int)
			bigInt.DivMod(bigInt, div, rem)
			if rem.Sign() != 0 {
				return 0, &IngressError{Code: "UNSAFE_JSON_NUMBER", Message: "number is not an integer"}
			}
		}

		if isNeg {
			bigInt.Neg(bigInt)
		}

		if !bigInt.IsInt64() {
			return 0, &IngressError{Code: "UNSAFE_JSON_NUMBER", Message: "number is out of safe integer range"}
		}

		val := bigInt.Int64()
		if val < MinSafeInteger || val > MaxSafeInteger {
			return 0, &IngressError{Code: "UNSAFE_JSON_NUMBER", Message: "number is out of safe integer range"}
		}
		return val, nil
	}

	parseValue = func(depth int) (any, error) {
		if depth > MaxJSONDepth {
			return nil, &IngressError{Code: "JSON_LIMIT_EXCEEDED", Message: fmt.Sprintf("nesting depth exceeds maximum %d", MaxJSONDepth)}
		}
		skipWhitespace()
		if pos >= len(text) {
			return nil, &IngressError{Code: "MALFORMED_JSON", Message: "unexpected end of JSON input"}
		}
		ch := text[pos]
		switch {
		case ch == '{':
			return parseObject(depth)
		case ch == '[':
			return parseArray(depth)
		case ch == '"':
			return parseString()
		case ch == 't':
			return parseLiteral("true", true)
		case ch == 'f':
			return parseLiteral("false", false)
		case ch == 'n':
			return parseLiteral("null", nil)
		case ch == '-' || (ch >= '0' && ch <= '9'):
			return parseNumber()
		default:
			return nil, &IngressError{Code: "MALFORMED_JSON", Message: "unexpected character"}
		}
	}

	parseObject = func(depth int) (map[string]any, error) {
		if err := countToken(); err != nil {
			return nil, err
		}
		pos++ // skip {
		obj := make(map[string]any)
		keys := make(map[string]bool)
		memberCount := 0

		skipWhitespace()
		if pos < len(text) && text[pos] == '}' {
			if err := countToken(); err != nil {
				return nil, err
			}
			pos++
			return obj, nil
		}

		for pos < len(text) {
			skipWhitespace()
			if text[pos] != '"' {
				return nil, &IngressError{Code: "MALFORMED_JSON", Message: "expected member name string"}
			}
			key, err := parseString()
			if err != nil {
				return nil, err
			}
			if keys[key] {
				return nil, &IngressError{Code: "DUPLICATE_JSON_MEMBER", Message: "duplicate object member name"}
			}
			keys[key] = true
			memberCount++
			if memberCount > MaxObjectMembers {
				return nil, &IngressError{Code: "JSON_LIMIT_EXCEEDED", Message: fmt.Sprintf("object member count exceeds maximum %d", MaxObjectMembers)}
			}

			skipWhitespace()
			if pos >= len(text) || text[pos] != ':' {
				return nil, &IngressError{Code: "MALFORMED_JSON", Message: "expected ':' after member name"}
			}
			if err := countToken(); err != nil {
				return nil, err
			}
			pos++ // skip :

			val, err := parseValue(depth + 1)
			if err != nil {
				return nil, err
			}
			obj[key] = val

			skipWhitespace()
			if pos >= len(text) {
				return nil, &IngressError{Code: "MALFORMED_JSON", Message: "unexpected end of object"}
			}
			nextChar := text[pos]
			if nextChar == '}' {
				if err := countToken(); err != nil {
					return nil, err
				}
				pos++
				return obj, nil
			}
			if nextChar == ',' {
				if err := countToken(); err != nil {
					return nil, err
				}
				pos++
			} else {
				return nil, &IngressError{Code: "MALFORMED_JSON", Message: "expected ',' or '}'"}
			}
		}
		return nil, &IngressError{Code: "MALFORMED_JSON", Message: "unterminated object literal"}
	}

	parseArray = func(depth int) ([]any, error) {
		if err := countToken(); err != nil {
			return nil, err
		}
		pos++ // skip [
		var arr []any

		skipWhitespace()
		if pos < len(text) && text[pos] == ']' {
			if err := countToken(); err != nil {
				return nil, err
			}
			pos++
			return arr, nil
		}

		for pos < len(text) {
			if len(arr) >= MaxArrayItems {
				return nil, &IngressError{Code: "JSON_LIMIT_EXCEEDED", Message: fmt.Sprintf("array item count exceeds maximum %d", MaxArrayItems)}
			}
			val, err := parseValue(depth + 1)
			if err != nil {
				return nil, err
			}
			arr = append(arr, val)

			skipWhitespace()
			if pos >= len(text) {
				return nil, &IngressError{Code: "MALFORMED_JSON", Message: "unexpected end of array"}
			}
			nextChar := text[pos]
			if nextChar == ']' {
				if err := countToken(); err != nil {
					return nil, err
				}
				pos++
				return arr, nil
			}
			if nextChar == ',' {
				if err := countToken(); err != nil {
					return nil, err
				}
				pos++
			} else {
				return nil, &IngressError{Code: "MALFORMED_JSON", Message: "expected ',' or ']'"}
			}
		}
		return nil, &IngressError{Code: "MALFORMED_JSON", Message: "unterminated array literal"}
	}

	rootValue, err := parseValue(1)
	if err != nil {
		return nil, err
	}
	skipWhitespace()
	if pos < len(text) {
		return nil, &IngressError{Code: "MALFORMED_JSON", Message: "unexpected trailing content"}
	}
	return rootValue, nil
}

func stringsHasPrefix(s, prefix string) bool {
	return len(s) >= len(prefix) && s[:len(prefix)] == prefix
}
